R/build_universe.R

```diff
@@ -13,24 +13,87 @@ build_universe <- function(input = getwd(), output = "packages.json") {
   assert_character_scalar(output, "invalid output")
   assert_file(input)
   packages <- list.files(input, all.files = FALSE, full.names = TRUE)
-  contents <- lapply(X = packages, FUN = read_package_entry)
-  out <- data.frame(
-    package = trimws(basename(packages)),
-    url = trimws(unlist(contents, use.names = FALSE)),
-    branch = "*release"
-  )
+  message("Processing ", length(packages), " package entries.")
+  entries <- lapply(X = packages, FUN = read_package_entry)
+  message("Aggregating ", length(entries), " package entries.")
+  aggregated <- do.call(what = vctrs::vec_rbind, args = entries)
   if (!file.exists(dirname(output))) {
     dir.create(dirname(output))
   }
-  jsonlite::write_json(x = out, path = output)
+  message("Writing packages.json.")
+  jsonlite::write_json(x = aggregated, path = output)
   invisible()
 }
 
 read_package_entry <- function(package) {
-  out <- readLines(con = package, warn = FALSE)
-  message <- assert_package(name = basename(package), url = out)
+  message("Processing package entry ", package)
+  name <- trimws(basename(package))
+  lines <- readLines(con = package, warn = FALSE)
+  out <- try(jsonlite::parse_json(lines), silent = TRUE)
+  if (inherits(out, "try-error")) {
+    package_entry_url(name = name, url = lines)
+  } else {
+    package_entry_json(name = name, json = out)
+  }
+}
+
+package_entry_url <- function(name, url) {
+  message <- assert_package(
+    name = name,
+    url = url,
+    assert_cran_url = FALSE # Prevents massive slowdown from 20000+ packages.
+  )
   if (!is.null(message)) {
     stop(message, call. = FALSE)
   }
-  out
+  data.frame(
+    package = trimws(name),
+    url = trimws(url),
+    branch = "*release"
+  )
+}
+
+package_entry_json <- function(name, json) {
+  fields <- names(json)
+  good_fields <- identical(
+    sort(fields),
+    sort(c("package", "url", "branch", "subdir"))
+  )
+  if (!good_fields) {
+    stop(
+      "Custom JSON entry for package ",
+      shQuote(name),
+      " must have fields 'packages', 'url', 'branch', and 'subdir' ",
```

Comment:
Can we have both? Otherwise, the subdir could be a development version, as is the case with `paws`.

Reply:
ah - hang on does the logic require all 4 fields? what if we want to submit a json file of normal packages without subdir?

Reply:
> what if we want to submit a json file of normal packages without subdir?

In that case, I think we would prefer the one-line URL format instead of JSON.

Reply:
sure, ok then.